### PR TITLE
Reject direct parameter entity recursion (part of #839)

### PR DESCRIPTION
**Pull request by hartwork**
Part of #839

CC (in alphabetic order) @catenacyber @RMJ10 @Snild-Sony @timbray

### CODE REVIEW

**Review by Snild-Sony — APPROVED** [review]
LGTM, but take my opinion with a grain of salt here as I haven't really looked too closely at this part of expat.

**Comment by hartwork** [comment]
@Snild-Sony thanks for having a look! :+1:

**Comment by RMJ10** [comment]
Looks good to me too.

**Comment by hartwork** [comment]
@RMJ10 thanks for the review!

File: expat/lib/xmlparse.c [hunk 1]
```diff
@@ -6240,7 +6240,7 @@ storeEntityValue(XML_Parser parser, const ENCODING *enc,
           dtd->keepProcessing = dtd->standalone;
           goto endEntityValue;
         }
-        if (entity->open) {
+        if (entity->open || entity == parser->m_declEntity) {
```
Comment (timbray) [inline expat/lib/xmlparse.c:6243]: If this is consistent with your coding style OK, but I always like to group conditions like this with parentheses, `(entity->open || (entity == parser->m_declEntity))`

**Review by timbray — APPROVED** [review]
I note that https://www.w3.org/TR/xml/#dt-PE specifically says that recursion of PE's isn't allowed.

File: expat/lib/xmlparse.c (hunk 1 above)
Comment (hartwork) [reply expat/lib/xmlparse.c]: @timbray alright, adjusted and pushed. Thanks for the review!

**Review by timbray — APPROVED** [review]

**Pull request merged**